### PR TITLE
Add Vuex support to `vue/no-unused-properties`

### PR DESCRIPTION
**Pull request by fuale**
Changes are aimed at improving Vuex support.

Following Vuex methods will be checked: `mapGetters`, `mapState`, `mapActions`, `mapMutations`, when `no-unused-properties` will be enabled

example:
```vue
<script>
  export default {
    computed: {
      ...mapGetters(["bar"]),
    },
    methods: {
      baz() {
        return this.bar // no error
      }
    }
  }
</script>
<template>
  {{ baz() }}
</template>
```

Changes will affect #1675 (Vuex and Pinia support)

### CODE REVIEW

**Comment by fuale** [comment]
The issue was created on October 22, 2021, but I am unable to find a way to check Vuex map methods today. To refactor my personal project, I am adding checks now. Maybe it would be useful to you

File: lib/rules/no-unused-properties.js [hunk 1]
```diff
@@ -453,8 +455,186 @@ module.exports = {
...
+        'CallExpression[callee.name=mapGetters][arguments] ObjectExpression Identifier[name]'(
+          node,
+          vueNode
+        ) {
+          const container = getVueComponentPropertiesContainer(vueNode.node)
+
+          container.properties.push({
+            type: 'array',
+            name: node.name,
+            groupName: 'props',
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:631]:
```suggestion
            groupName: 'computed',
```

File: tests/lib/rules/no-unused-properties.js
Comment (FloEdelmann) [inline tests/lib/rules/no-unused-properties.js:1]: Some test case is duplicated, see the lint error. Please remove one of the duplicated test cases (you can run `npm run lint` locally).

File: lib/rules/no-unused-properties.js [hunk 2]
```diff
@@ -453,8 +455,186 @@ module.exports = {
...
+            add: 'increment'
+          })
+
+          methods: {
+            ...mapMutations({
+              add: 'increment'
+            })
+          }
+        */
+        'CallExpression[callee.name=mapMutations][arguments] ObjectExpression Property[key.name]'(
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:469]: Can we somehow combine and simplify these cases? At lease object and array expression could be handled in the same visitor, maybe even all Vuex function calls.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Sorry for the late review. Thanks for your pull request! I have some comments, see below.

**Comment by FloEdelmann** [comment]
@fuale Would you fancy implementing the review feedback?

File: lib/rules/no-unused-properties.js (hunk 2 above)
Comment (fuale) [reply lib/rules/no-unused-properties.js]: It is now better?

File: lib/rules/no-unused-properties.js [hunk 3]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        /*
+          mapActions or mapMutations.
+
+          methods: mapMutations({ add: 'increment' })
+
+          methods: {
+            ...mapMutations({ add: 'increment' })
+          }
+        */
+        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ObjectExpression Property[key.name]'(
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:467]:
```suggestion
        /**
         * e.g. `mapMutations({ add: 'increment' })`
         * @param {Property} node
         * @param {VueObjectData} vueData
         */
        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ObjectExpression Property'(
```

File: lib/rules/no-unused-properties.js [hunk 4]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        /*
+          mapActions or mapMutations.
+
+          methods: mapMutations(['add'])
+
+          methods: {
+            ...mapMutations(['add'])
+          }
+        */
+        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ArrayExpression Literal[value]'(
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:490]:
```suggestion
        /**
         * e.g. `mapMutations(['add'])`
         * @param {Literal} node
         * @param {VueObjectData} vueData
         */
        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ArrayExpression Literal'(
```

File: lib/rules/no-unused-properties.js [hunk 5]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+            })
+          }
+
+          computed: mapState({
+            count (state) {
+              return state.todosCount
+            }
+          })
+        */
+        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ObjectExpression Property[key.name]'(
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:523]:
```suggestion
        /**
         * e.g. `mapState({ count: state => state.todosCount })`
         * @param {Property} node
         * @param {VueObjectData} vueData
         */
        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ObjectExpression Property'(
```

File: lib/rules/no-unused-properties.js [hunk 6]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        */
+        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ObjectExpression Property[key.name]'(
+          node,
+          vueNode
+        ) {
+          const container = getVueComponentPropertiesContainer(vueNode.node)
+
+          container.properties.push({
+            type: 'array',
+            name: node.key.name,
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:475]:
```suggestion
            name: utils.getStaticPropertyName(node),
```

File: lib/rules/no-unused-properties.js [hunk 7]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        */
+        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ObjectExpression Property[key.name]'(
+          node,
+          vueNode
+        ) {
+          const container = getVueComponentPropertiesContainer(vueNode.node)
+
+          container.properties.push({
+            type: 'array',
+            name: node.key.name,
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:531]:
```suggestion
            name: utils.getStaticPropertyName(node),
```

File: lib/rules/no-unused-properties.js [hunk 8]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        /*
+          mapState or mapGetters.
+
+          computed: mapGetters(['count1', 'count2'])
+
+          computed: {
+            ...mapGetters(['count']),
+          }
+        */
+        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ArrayExpression Literal[value]'(
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:546]:
```suggestion
        /**
         * e.g. `mapState(['count'])`
         * @param {Literal} node
         * @param {VueObjectData} vueData
         */
        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ArrayExpression Literal'(
```

File: lib/rules/no-unused-properties.js [hunk 9]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        */
+        'CallExpression[callee.name=/^(mapMutations|mapActions)$/][arguments] ArrayExpression Literal[value]'(
+          node,
+          vueNode
+        ) {
+          const container = getVueComponentPropertiesContainer(vueNode.node)
+
+          container.properties.push({
+            type: 'array',
+            name: node.value,
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:498]:
```suggestion
            name: utils.getStringLiteralValue(node),
```

File: lib/rules/no-unused-properties.js [hunk 10]
```diff
@@ -453,8 +455,110 @@ module.exports = {
...
+        */
+        'CallExpression[callee.name=/^(mapState|mapGetters)$/][arguments] ArrayExpression Literal[value]'(
+          node,
+          vueNode
+        ) {
+          const container = getVueComponentPropertiesContainer(vueNode.node)
+
+          container.properties.push({
+            type: 'array',
+            name: node.value,
```
Comment (FloEdelmann) [inline lib/rules/no-unused-properties.js:554]:
```suggestion
            name: utils.getStringLiteralValue(node),
```

**Review by FloEdelmann — COMMENTED** [review]
Some suggestions. Please try them out locally and see if there are any type conflicts now.

**Comment by CyanSalt** [comment]
It is worth mentioning that Pinia also provides `mapWritableState`, which may also produce properties.

For reference, `eslint-plugin-galaxy` created by me once implemented [similar functionality](https://github.com/CyanSalt/eslint-plugin-galaxy/blob/main/src/vue/no-unused-vuex-properties.ts) (although it may have issues).

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you! Sorry for the late review.

**Comment by eigan** [comment]
@fuale Thanks for this! Would it be easy to also add support for Pinia `mapStores`? https://pinia.vuejs.org/api/modules/pinia.html#mapStores 👼🏻

**Comment by FloEdelmann** [comment]
I'm afraid that's not possible, since the name of the resulting properties can't be easily deduced.

**Comment by eigan** [comment]
@FloEdelmann allright, thanks!

**Comment by luizotcarvalho** [comment]
I really appreciate this feature, thanks @fuale ! While using I found a edge case, Im using Nuxt 2 and mapping a Pinia action to my page but this action is called with a `vm` inside the `beforeRouteEnter` like this:

```js
methods: {
  ...mapActions(useStatementStore, {
    resetStatement: '$reset'
  }),
},
...
beforeRouteEnter (to, from, next) {
    next(vm => vm.resetStatement())
}
```

I keep receiving the error `'resetStatement' of method found, but never used` and don't know how to proceed, you think will be need to ignore cases like this or the support can be broaden a little?

**Comment by FloEdelmann** [comment]
@luizotcarvalho please create a new issue following the bug template, so this can be addressed properly.

**Comment by Ceall8650** [comment]
I've upgraded my `eslint-plugin-vue` plugin to the version: `9.24.0`
but the unused `mapState` still doesn't get the error. 
Should I need to upgrade other plugins or some necessary plugins that I need to install?

Here is my `package.json`
```
// package.json
...
"vue": "^2.6.11",
"vue-eslint-parser": "^9.4.2",
"eslint-plugin-vue": "^9.24.0",
```

```js
// .eslintrc.js
module.exports = {
  ...
  parser: "vue-eslint-parser",
  parserOptions: {
    parser: '@babel/eslint-parser',
  },
  env: {
    browser: true,
  },
  extends: [
    'plugin:vue/recommended',
    'airbnb-base'
  ],
  plugins: [
    'vue',
    'import',
    'import-helpers'
  ],
  rules: {
    ...,
    "vue/no-unused-properties": ["error", {
      "groups": [
        'props',
        'computed',
        'data',
        'methods',
      ],
      "deepData": true,
    }]
}
```

**Comment by FloEdelmann** [comment]
Please open a new issue with your exact setup and sample code. But before that, please try upgrading your ESLint version, you're a major version behind there.

**Comment by Ceall8650** [comment]
OK, Thanks

**Pull request merged**